Feature/semi step2

semi-cloud opened the pull request:
[FileMetaData]

- 폴더도 하나의 파일이기 때문에 굳이 파일과 폴더 테이블을 분리하지 않고, `FileType` 칼럼을 통해서 동일 테이블 내에서 구분해줬습니다. 
  + 폴더의 `uploadName` 에는 `.` 를 저장하도록 했습니다. 해당 필드를 통해 파일/폴더를 구분할 수 있지만, `enum` 이 명시적으로 더 잘 구분되는 것 같아서 생성했습니다.
- 처음에는 `folderPath` 라는 필드를 두어 현재 파일/폴더가 속해있는 부모 폴더의 논리적인 경로를 저장했지만, 굳이 파일 서버에 계층 구조를 유지할게 아니라면 아래와 같은 이유로 경로를 추적하지 않는게 낫다고 판단했습니다.
    + 파일의 논리적인 경로를 추적하면 폴더 이름 수정, 폴더 이동 시 재귀적으로 하위에 있는 모든 파일 및 폴더의 경로 또한 변경해줘야 하므로 큰 부하가 걸립니다.
    + 파일 정보를 보여주는 API에서 파일의 논리적인 경로가 필요하지만, 요청 자체가 많지 않을 것 같고 요청이 오더라도 재귀적으로 위로 올라가는 방향이기 때문에 밑으로 타고 내려가는 비용보다 적을 것 같다 판단했습니다. (혹은 프론트에서 처리)
    + 예를 들어 현재 파일이 깊이 10 위치에 저장되어 있고, 뒤에 만개의 파일이 존재한다 했을 때 해당 폴더 이름 변경 시 만개의 파일의 경로 모두를 수정해줘야 하지만 논리적인 경로만 가져온다면 위로 10번만 올라가면 되기 때문입니다. (근데 지금 생각해보니 `bulkupdate` 하면 부하를 좀 줄일 수 있겠네요.)
    
[FolderService]

- 각 사용자에 해당하는 루트 폴더를 하나씩 만들어주었습니다.
    - 파일과 마찬가지로, 물리적인 폴더 저장이나 DB 메타데이터 저장이 실패해 롤백되는 상황을 고려해 1. 중복 검사 2. 메타데이터 저장 3. 물리적인 폴더 저장 순으로 구현했습니다.
    - 루트 폴더 생성할 때 `parent` 에 `null` 값이 들어가는데 혹시나 문제될게 있을지 고민됩니다.
- 루트 폴더를 생성한 이후에는, 계층 구조로 보여주는 것은 프론트의 몫이라고 생각해 굳이 물리적인 서브 폴더를 생성하지 않도록 했습니다.
    - 실제 파일 서버에 계층 구조를 유지하면 파일 I/O 로 인한 부하가 클 것이라 생각했습니다.
- 파일의 논리적인 경로가 있었다면 변경하려는 폴더 아래에 있는 모든 자식들의 경로도 변경이 필요했겠지만, 없앰에 따라 자식들은 제외하고 변경하려는 폴더 자체만 이름 수정하도록 했습니다.

[FileMetaDataReader, FileMetaDataWriter]

- 엔티티를 생성하고 저장하는 로직이 중복되어 데이터베이스 CRUD 로직을 따로 분리했습니다.
- 메서드를 따로 추출하고 내부에 다시 빌더를 쓰는게 좀 요상한것 같기도 하네요.

[FileService]

- 파일은 모두 사용자 별 루트 폴더 내부에 저장되어야 하도록 수정했습니다.
- 폴더가 생김에 따라, 파일 중복 여부를 검사할 때 같은 폴더 아래에 파일이 존재하는지 확인하도록 수정했습니다.

semi-cloud committed:
feat: 폴더 저장 구현

[FileMetaData]

- 폴더도 하나의 파일이기 때문에 굳이 파일과 폴더 테이블을 분리하지 않고, FileType 칼럼을 통해서 동일 테이블 내에서 구분해줬습니다.
- 해당 폴더에 해당하는 파일 목록을 객체 지향적인 코드로 가져올 수 있도록, 양방향 연관관계 메서드를 추가했습니다.
- folderPath 에는 현재 파일/폴더가 속해있는 부모 폴더의 논리적인 경로를 저장합니다.
    - DB에는 실제 물리 경로까지 저장할 필요가 없다고 판단했습니다.
- 일단 폴더의 uploadName에는 “.” 를 저장하도록 했습니다.
    - enum 컬럼을 따로 두지 않고도 UploadName으로 구분할 수 있지만, enum이 명시적으로 더 잘 구분되는 것 같아서 생성했습니다.

[PathService]

- 파일 논리 경로 생성 : folderPath + uploadName ex) /semi/folder1 + / + image.png
- 파일 물리 경로 생성 : BaseDir + folderPath + uploadName ex) C://.. + /semi/folder1 + / + image.png

[FolderService]

- 각 사용자에 해당하는 루트 폴더를 하나씩 만들어주었습니다.
    - 파일과 마찬가지로, 물리적인 폴더 저장이나 DB 메타데이터 저장이 실패해 롤백되는 상황을 고려해 1. 중복 검사 2. 메타데이터 저장 3. 물리적인 폴더 저장 순으로 구현했습니다.
    - 이런 경우 루트 폴더이므로 부모 칼럼 값으로 null을 가지게 되는데, 괜찮은지 모르겠네요.
- 루트 폴더를 생성한 이후에는, 계층 구조로 보여주는 것은 프론트의 몫이라고 생각해 굳이 물리적인 서브 폴더를 생성하지 않도록 했습니다.

[FileMetaDataReader, FileMetaDataWriter]

- 엔티티를 생성하고 저장하는 로직이 중복되어 데이터베이스 CRUD 로직을 따로 분리했습니다.

semi-cloud committed:
feat: 파일 이름 수정 기능 구현

[FileMetaData]

- FolderPath 제거
    - 파일의 논리적인 경로를 추적하면 폴더 이름 수정, 폴더 이동 시 재귀적으로 하위에 있는 모든 파일 및 폴더의 경로 또한 변경해줘야 하므로 너무 큰 부하가 걸린다 생각해 제거했습니다.
    - 파일 정보를 보여주는 API에서 해당 파일의 논리적인 경로가 필요하지만, 요청 자체가 많지 않을 것 같고 요청이 오더라도 재귀적으로 위로 올라가는 방향이기 때문에 밑으로 타고 내려가는 비용보다 적을 것 같다 판단했습니다.
        - 예를 들어 현재 파일이 깊이 10 위치에 저장되어 있고, 뒤에 만개의 파일이 존재한다 했을 때 해당 폴더 이름 변경 시 밑에 만개의 파일  모두 경로를 수정해줘야 하지만 논리적인 경로만 가져온다면 위로 10번만 올라가면 되기 때문입니다.

[FileService]

- 파일은 모두 사용자 별 루트 폴더 내부에 저장되어야 하도록 수정했습니다.
- 폴더가 생김에 따라, 파일 중복 여부를 검사할 때 같은 폴더 아래에 파일이 존재하는지 확인하도록 수정했습니다.

[FolderService]

- 파일의 논리적인 경로가 있었다면 변경하려는 폴더 아래에 있는 모든 자식들의 경로도 변경이 필요했겠지만, 없앰에 따라 자식들은 제외하고 변경하려는 폴더 자체만 이름 수정하도록 했습니다.
- 부모 폴더 생성할 때 parent에 null 값이 들어가는데 혹시나 문제될게 있을지 고민됩니다.

[PathService]

- FolderPath 삭제에 따라 논리적인 경로 반환하는 메서드를 제거했습니다.

[FileMetatDataReader]

- 중복 검사 등 예외를 발생하는 validate 메서드 자체를 데이터 계층에 넣었는데, 기존대로 서비스 계층에 존재하는게 나을지 고민됩니다.

semi-cloud committed:
feat: 폴더 조회 기능 구현

semi-cloud committed:
test: 폴더 컨트롤러 테스트

semi-cloud committed:
test: 서비스 및 데이터 계층 단위 테스트 작성

[FileMetaDataWriterTest]
- Builder에 parnet 필드와 스터빙해준 값 저장이 안되는 이슈 발생
- 우선 @SpringBootTest로 실제 동작 검증 대신 verify로 메서드 실행 여부만 체크하도록 수정

semi-cloud committed:
test: FileUtil 폴더 테스트 추가

semi-cloud committed:
refactor: 코드 스멜 제거

pear96 commented on src/main/java/com/c4cometrue/mystorage/api/dto/FolderGetDto.java:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	public record FileDto(
+		long fileId,
```
`Long` 과 `long`의 차이를 의도적으로 둔 건지 궁금합니다.

pear96 commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataReader.java:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	public boolean isDuplicateFile(String fileName, long userId, FileMetaData parent) {
+		return repository.existsByFileNameAndUserIdAndParent(fileName, userId, parent);
```
오 중복 검색은 그냥 `existsBy`가 나을 것 같네요 🤓

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/api/dto/FolderGetDto.java:
```diff
@@ -0,0 +1,44 @@
+public class FolderGetDto {
+
+	public record Req(
+		@NotNull(message = "유저 ID는 Null이 될 수 없습니다.") Long userId,
```
NotNull 이면 long 쓰는 건 어떤 가요

semi-cloud replied:
@kochungcheon 
저도 처음에 그렇게 했었는데 Bind 관련 Exception이 터지려면 애초에 Long으로 null값을 허용해줘야 하더라구요!
ExceptionHandler에서 예외 처리 한꺼번에 처리해주고 싶어서 Long 으로 받고, 서비스 단부터는 원시 타입 사용했습니다.

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/api/FolderController.java:
```diff
@@ -0,0 +1,59 @@
+
+	private final FolderService folderService;
+
+	@PostMapping("/root")
```
오 좋네요

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/domain/FileMetaData.java:
```diff
@@ -2,44 +2,76 @@
 	@Column(name = "file_meta_data_id")
 	private Long id;
 
+	@ManyToOne(fetch = FetchType.LAZY)
```
FetchType.LAZY 쓰신 이유가 궁금합니다!

semi-cloud replied:
특정 폴더나 파일 정보를 가져올때, 부모 폴더 정보까지 한꺼번에 가져오는 로직이 없다고 판단되어 지연 로딩 전략 사용했습니다.

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/domain/FileMetaData.java:
```diff
@@ -2,44 +2,76 @@
+		this.parent = parent;
+	}
+
+	public void rename(String newName) {
```
불변 객체를 고려해보는 방법도 있을 거 같아요

semi-cloud replied:
@kochungcheon 
음..👀 잘 이해가 안가는데 혹시 예시 들어주실 수 있나요?

kochungcheon replied:
> @kochungcheon 음..👀 잘 이해가 안가는데 혹시 예시 들어주실 수 있나요?

불변객체 고려한다면
새로운 객체 생성하면 될거같아요!
객체 내부 값을 바꾸는게 아니라

semi-cloud replied:
아하 ㅎㅎ JPA 더티 체킹 기능 사용하려 해서 내부 값을 바꿨는데, 불변도 좋을 것 같네요.

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataRepository.java:
```diff
@@ -1,11 +1,20 @@
 import com.c4cometrue.mystorage.domain.FileMetaData;
 
 public interface FileMetaDataRepository extends JpaRepository<FileMetaData, Long> {
 
```
적절한 인덱스 전략을 고려해보면 좋을 거 같아요!

semi-cloud replied:
Step7 세팅 끝나고 테스트해보면서 인덱스 생성하려 했는데 그냥 지금 해야겠네요 굳굳 👍

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -25,48 +23,46 @@
-		String uploadFilePath = getFullUploadFilePath(originName);
-		FileMetaData fileMetaData = saveFileMetaData(file, userId, uploadFilePath);
+		// 파일 메타데이터 저장
+		String uploadFileName = pathService.createUniqueFileName(originName);
```
해당 유니크 네임이 정말 유니크한 지 검증하는 로직을 추가하는 건 어떤가요

semi-cloud replied:
`UUID.randomUUID()` 로 생성하고 있는데 다시 한번 검증하라는 말씀이실까요!?

VSFe replied:
randomUUID() 라고 해서 무조건적으로 안겹칠거라는 보장은 없으니까요.

kochungcheon reviewed (commented):
고생하셨습니다!

pear96 commented on src/main/java/com/c4cometrue/mystorage/domain/FileMetaData.java:
```diff
@@ -2,44 +2,76 @@
 @Entity
 @Getter
-@NoArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
무지성으로 `@RequiredArgsConstructor`를 사용했는데 찾아보는 계기가 되었습니다.
JPA는 기본 생성자를 요구하기 때문에 `@NoArgsConstructor`가 더 바람직하다고 하는군요.
참고로 final이나 `@NonNull`이 있으면 `@RequiredArgsConstructor`를 써야한다고 하는군요.

pear96 commented on src/main/java/com/c4cometrue/mystorage/domain/FileMetaData.java:
```diff
@@ -2,44 +2,76 @@
+	@JoinColumn(name = "parent_id")
+	private FileMetaData parent;
+
 	@Column(nullable = false)
```
`@NotNull`대신 사용한 이유가 있을까요? 둘의 차이점을 아시나요?

semi-cloud replied:
`@NotNull` 은 하이버네이트 단에서 예외가 발생하고, `@Column(nullable = false)` 은 쿼리 나가고 DB 단에서 예외가 발생할 것 같네요. 궁금해서 찾아봤는데 JPA 에서 `ConstraintViolationException` 을 발생시켜서 @Valid 가 없어도 엔티티를 검증할 수 있어 더 안전하네요 감사합니다 👍

정리해놓은 좋은 글이 있어서 같이 첨부할게요.
https://kafcamus.tistory.com/15

pear96 commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -78,42 +74,18 @@ public FileDownloadDto.Response fileDownLoad(long userId, long fileId) {
-			() -> new BusinessException(ErrorCode.FILE_NOT_FOUND));
+	private FileMetaData getParentFolder(FileMetaData root, long userId, long folderId) {
+		if (root.getId() != folderId) {
+			fileMetaDataReader.get(folderId, userId);
```
루트가 아닌 부모 폴더를 찾기만 하는건가요? 할당되는 부분이 안보여서요..
제가 보기엔 root만 반환될 것 같은데, 잘못 본걸까요 @.@

semi-cloud replied:
엇 리턴문이 어디갔지..🥹

pear96 commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,95 @@
+		FileMetaData folder = getFolder(folderId, userId);
+
+		// 폴더 내부에 존재하는 파일 및 폴더 목록 반환
+		List<FileMetaData> files = folderReader.getFiles(userId, folder);
```
userId가 필요한지 저는 좀 확신이 안서네요

semi-cloud replied:
음 로그인 API 가 있었다면 `folderId` 로만 조회하는게 맞는데,
현재 시큐리티 관련해서 설정해놓은게 없어서 더블 체크 차원에서 `userId` 까지 추가한거라, 굳이 없어도 될것 같긴 합니닷.. 👀

pear96 commented on src/main/java/com/c4cometrue/mystorage/utils/FileUtil.java:
```diff
@@ -56,4 +57,16 @@ public void deleteFile(String fileUploadPath) {
+
+		try {
+			Files.createDirectory(destinationPath);
+		} catch (FileAlreadyExistsException ex) {
```
오 저도 다른 예외를 처리해야겠네요. 공식문서를 보니 createDirectory 말고도 다른 메서드도 처리해야할 예외가 많군요.. 😂

pear96 reviewed (commented):
File과 Folder를 합친것, Reader와 Writer를 나눈 것이 인상깊었습니다. 청천도 그랬던걸 생각하면 그게 보편적인건가 싶은 생각이 드네요.
세미님 코드 보면서 저도 많이 배웠습니다! 더 추가할 내용이 있다면 추가하겠습니다. 한번에 다 읽기가 쉽지 않네요

VSFe commented on src/main/java/com/c4cometrue/mystorage/api/FileController.java:
```diff
@@ -31,9 +32,9 @@ public class FileController {
 	@ResponseStatus(HttpStatus.CREATED)
 	public FileUploadDto.Response upload(
-		@RequestPart MultipartFile file, long userId
+		@RequestPart @NotNull MultipartFile file, long userId, long folderId
```
기왕 multiPartFile에 validation을 건 겸, userId나 folderId에도 하나 만들어주죠.
`@Positive` 정도가 적당하겠네요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/api/FolderController.java:
```diff
@@ -0,0 +1,59 @@
+	private final FolderService folderService;
+
+	@PostMapping("/root")
+	@ResponseStatus(HttpStatus.CREATED)
```
단 한번만 실행되어야 하는 API면, 차라리 register 관련 api를 정의해서 해당 api 내부에서 루트 폴더를 만드는게 좀 더 자연스러울 것 같아요.

또한, 루트 폴더에 폴더 이름이 있어야 하나요?

semi-cloud replied:
아하.. 사용자 이름으로 생성하려 했는데 굳이 루트 폴더에 이름이 없어도 되긴 하겠군요 👍
그러면 "." 도 아니고 그냥 `null` 로 놔두는게 나을까요?

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/PathService.java:
```diff
@@ -0,0 +1,35 @@
+	private String baseDir;
+
+	public String createUniqueFileName(String originFileName) {
+		return UUID.randomUUID() + originFileName;
```
originFileName이 필요한 이유가 있을지 잘 모르겠네요.

semi-cloud replied:
하핫 그렇네요
지금 보니 `UUID()` 가 안전함을 100% 보장할 수 없다면, 중복 가능성이 제일 높은 파일 이름보다는 업로드 날짜(밀리초까지?)를 추가하는게 나을것 같습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,95 @@
+import lombok.RequiredArgsConstructor;
+
+@Service
+@Transactional
```
최상위 `@Transactional` 은 자제하는게 좋습니다.

semi-cloud replied:
@VSFe 
헉 혹시 이유를 여쭤봐도 될까요? 가독성 측면 때문인가..?

VSFe commented on src/main/java/com/c4cometrue/mystorage/common/exception/ErrorCode.java:
```diff
@@ -24,7 +25,12 @@ public enum ErrorCode {
 
 	// file-delete
-	FILE_DELETE_FAILED(HttpStatus.INTERNAL_SERVER_ERROR, "파일 삭제에 실패했습니다.");
+	FILE_DELETE_FAILED(HttpStatus.INTERNAL_SERVER_ERROR, "파일 삭제에 실패했습니다."),
```
구현한 API가 별로 없는데, 에러 코드가 너무 많아지는 기분이네요.
너무 촘촘히 구성할 필요는 없어요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/domain/FileMetaData.java:
```diff
@@ -2,44 +2,76 @@
 
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "parent_id")
+	private FileMetaData parent;
```
Self Join을 사용하고 있는데, toString() 이나 hashCode() 사용할 일 있으면 꼭!!!!! 재 확인하세요.
서비스 장애의 원인이 될 수도 있습니다.

semi-cloud replied:
오 전혀 생각하지 못했던 부분이네요..👀
흐음 혹시 무한 재귀 현상으로 인한 스택 오버플로우가 원인이 맞을까요!?

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataReader.java:
```diff
@@ -0,0 +1,37 @@
+
+import lombok.RequiredArgsConstructor;
+
+@Component
```
의미론적으론 `@Repository` 아닐까요?

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataReader.java:
```diff
@@ -0,0 +1,37 @@
+
+	private final FileMetaDataRepository repository;
+
+	public FileMetaData get(long id, long userId) {
```
folder를 해당 메서드로 접근했는데 없으면 FILE_NOT_FOUND를 반환하겠네요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataReader.java:
```diff
@@ -0,0 +1,37 @@
+		return repository.existsByFileNameAndUserIdAndParent(fileName, userId, parent);
+	}
+
+	public List<FileMetaData> getFiles(long userId, FileMetaData parent) {
```
하위 파일이 한 4천개 있어도 전부 다 가져오겠네요.

semi-cloud replied:
아앗 페이징을 깜빡했네요 굳굳 👍

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataRepository.java:
```diff
@@ -1,11 +1,20 @@
+
+	Optional<FileMetaData> findByIdAndUserId(long fileId, long userId);
+
+	Optional<FileMetaData> findByUserIdAndParent(long userId, FileMetaData parent);
```
결과가 Unique 하지 않을 위험이 매우 높아요.

semi-cloud replied:
@VSFe 
으음..루트 폴더 찾을 때 사용되는 메서드인데, 유저별 루트 폴더는 하나만 존재하므로 유니크할 것이라 생각했습니다.
그래서 아무리 생각해봐도 모르겠네요 ㅠㅠ 어떤 상황에서 결과가 유니크 하지 않을까요..? 😭

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataWriter.java:
```diff
@@ -0,0 +1,52 @@
+
+@Component
+@RequiredArgsConstructor
+@Transactional
```
최상단 `@Transactional` 은 자제 부탁드릴게요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataWriter.java:
```diff
@@ -0,0 +1,52 @@
+		FileMetaData folderMetaData = FileMetaData.fileBuilder()
+			.userId(userId)
+			.fileName(folderName)
+			.uploadName(".")
```
파일 이름이 "." 이면 어떻게 하는게 좋을까요?

semi-cloud replied:
음..당장 이름으로 데이터를 찾는 로직은 없어서 둘이 `.` 로 같아도 될 것 같긴한데..@@
폴더 `uploadName` 도 `UUID()` 로 생성하는게 나을까요?

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -25,48 +23,46 @@
-	public FileUploadDto.Response fileUpload(MultipartFile file, long userId) {
-		if (Objects.isNull(file) || file.isEmpty()) {
+	public FileUploadDto.Response fileUpload(MultipartFile file, long userId, long folderId) {
+		if (file.isEmpty()) {
```
file Size가 0인 파일도 존재할 수 있지 않나요?
(ex. 빈 텍스트 파일)

semi-cloud replied:
해당 메서드 들어가보니까 아래처럼 `사이즈 = 0` 로 비교하고 있더라구요..!
```java
@Override
public boolean isEmpty() {
    return (this.part.getSize() == 0);
}
```

VSFe commented on src/main/java/com/c4cometrue/mystorage/utils/FileUtil.java:
```diff
@@ -56,4 +57,16 @@ public void deleteFile(String fileUploadPath) {
+			Files.createDirectory(destinationPath);
+		} catch (FileAlreadyExistsException ex) {
+			throw new BusinessException(ErrorCode.DUPLICATE_FOLDER, ex);
+		} catch (IOException ex) {
```
그 이외의 에러는 그대로 throw 되겠네요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/PathService.java:
```diff
@@ -0,0 +1,35 @@
+	 */
+	public String getFullFilePath(String parentDir, String fileName) {
+		String validatedFileName = fileName.trim();
+		if (Objects.equals(parentDir, "")) {
```
`StringUtils.isEmpty(parentDir)`

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/PathService.java:
```diff
@@ -0,0 +1,35 @@
+	 * @return
+	 */
+	public String getFullFilePath(String parentDir, String fileName) {
+		String validatedFileName = fileName.trim();
```
<img width="348" alt="image" src="https://github.com/C4-ComeTrue/my-storage/assets/4595546/025fdb38-cd33-4d39-87b2-a8c07544267e">

trim은 **앞**, 뒤 공백을 제거합니다.
앞쪽 공백은 문제가 없어요.

semi-cloud replied:
아 앞뒤 모두 공백 있으면 안될 것 같아서 싸그리 지웠는데 앞쪽은 살려둬야겠군요..👍

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -25,48 +23,46 @@
 
-		fileUtil.uploadFile(file, uploadFilePath);
+		// 실제 물리 파일 저장
+		String uploadFullPath = pathService.getFullFilePath(rootFolder.getFileName(), uploadFileName);
```
전체적인 로직을 보면, `rootFolder.getFileName()` 은 항상 "" 아닌가요?

semi-cloud replied:
네 근데 약간 하드코딩(?) 느낌인것 같아서 일관성 있게 네이밍은 메서드로 가져오도록 했습니다 🥹

VSFe commented on src/main/java/com/c4cometrue/mystorage/api/FolderController.java:
```diff
@@ -0,0 +1,59 @@
+
+	@PostMapping
+	@ResponseStatus(HttpStatus.CREATED)
+	public FolderUploadDto.Res createFolder(
```
사용자가 createRootFolder를 호출하지 않고, createFolder를 호출할 때 parentNo = null로 던지면 똑같은 결과가 나오겠네요.

semi-cloud replied:
`@NotNull` 로직을 추가해야겠네요!

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -25,48 +23,46 @@
-	 * @return fileContentByteArray, FileContentType
-	 */
 	@Transactional(readOnly = true)
 	public FileDownloadDto.Response fileDownLoad(long userId, long fileId) {
```
파일, 폴더를 하나의 테이블로 관리하려고 하다보니, `fileDownload`, `fileDelete` 와 같은 파일 전용 메서드에서 fileId에 folderId를 넣어도 문제 없이 동작 가능할 것으로 보입니다.

폴더 삭제를 시도 할 경우 하위 파일의 메타 데이터는 남아있지만 실제 폴더 데이터는 모두 지워지는 Dangling 상황이 발생할 위험이 너무 높아보이니, 방어로직을 개발하던 테이블 스키마를 변경하는 식의 대응이 필요합니다.

VSFe reviewed (changes requested):
우선, 반영하고 추가로 리뷰할 부분 있는지 확인해 볼게요.

semi-cloud committed:
refactor: 1차 피드백 반영

[Dto]
- ID에 @Positive 검증 추가

[FolderController]
- 루트 폴더 생성 작업은 레지스터 관련 API로 분리

[PathService]
- 파일명 UUID() 의 안전성을 올리기 위해 이름 대신 시간 데이터 추가

- 파일 원본명 뒤쪽 공백만 제거

- StringUtils.isEmpty deprecated 이슈로 StringUtils.hasLength로 수정

[FileService]
- ID에 해당하는 데이터가 파일/폴더인지 확인하기 위해 검증 로직 추가

semi-cloud commented:
@VSFe 
우선 질문 없이 바로 수정할 수 있었던 것부터 해결했습니다! 확인 부탁드려용
